### PR TITLE
Parseable report file

### PR DESCRIPTION
**Pull request by stalar**
This is a first shot at implementing https://github.com/google/error-prone/issues/444. Comments and feedback welcome!

This PR adds a new option called `-XepReportFile:<path>`. If this option is present, an xml document will be created in the location specified. The contents of the document looks like:

    <?xml version="1.0" encoding="utf-8"?>
    <errorprone timestamp="2018-10-02T13:05:37.389+02:00[Europe/Stockholm]">
    <descriptions>
      <description
        type="WARNING"
        source="/Users/sla/dev/example/src/main/java/com/example/Example.java"
        line="6"
        column="8">
        <message>[PrivateConstructorForUtilityClass] Utility classes (only static members) are not designed to be instantiated and should be made noninstantiable with a default constructor.
    (see https://errorprone.info/bugpattern/PrivateConstructorForUtilityClass)
    Did you mean to remove this line?</message>
      </description>
      ...
    </errorprone>

The implementation is done in a new class called `FileReporter` which is created on startup and passed down to the `JavacErrorDescriptionListener`. The class is also a `TaskListener` so that the file can be closed when the compilation is done.

### CODE REVIEW

**Comment by jbnerd** [comment]
Any possibilities of this getting merged any time soon?

**Comment by tomasbjerre** [comment]
@stalar Are you aware of the merge conflict? Really hope you get this merged =) Nice job!

**Comment by davidhart82** [comment]
+1 for this, we're having to parse an awfully large build output to find the errors. This should bring it on line with how every other static analysis tool works.

**Comment by davidhart82** [comment]
Though have you thought about reusing one of the other already supported formats available? It would mean that most tools, and in particular, CI systems would already be able to handle this.

File: check_api/src/main/java/com/google/errorprone/JavacErrorDescriptionListener.java [hunk 1]
```diff
@@ -57,11 +58,14 @@
 
   private JavacErrorDescriptionListener(
       Log log,
+      FileReporter fileReporter,
```
Comment (davidhart82) [inline check_api/src/main/java/com/google/errorprone/JavacErrorDescriptionListener.java]: Feels like we'd be better having a Reporter interface?

File: check_api/src/main/java/com/google/errorprone/FileReporter.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.time.ZonedDateTime;
+import org.apache.commons.text.StringEscapeUtils;
+
+/**
+ * Generates a parseable report file.
+ */
+public class FileReporter implements TaskListener {
```
Comment (davidhart82) [inline check_api/src/main/java/com/google/errorprone/FileReporter.java]: XMLFileReporter?

File: check_api/src/main/java/com/google/errorprone/FileReporter.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+      }
+    }
+  }
+
+  public void report(final JCDiagnostic.DiagnosticType type, final DiagnosticSource source,
+                     final JCDiagnostic.DiagnosticPosition pos, final String message) {
+    if (pw != null) {
+      pw.println("    <description");
+      pw.println("      type=\"" + type + "\"");
+      pw.println("      source=\"" + source.getFile().getName() + "\"");
```
Comment (davidhart82) [inline check_api/src/main/java/com/google/errorprone/FileReporter.java]: Probably want to escape this?

File: check_api/src/main/java/com/google/errorprone/FileReporter.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        pw.close();
+      }
+    }
+  }
+
+  public void report(final JCDiagnostic.DiagnosticType type, final DiagnosticSource source,
+                     final JCDiagnostic.DiagnosticPosition pos, final String message) {
+    if (pw != null) {
+      pw.println("    <description");
+      pw.println("      type=\"" + type + "\"");
```
Comment (davidhart82) [inline check_api/src/main/java/com/google/errorprone/FileReporter.java]: Probably want to escape this?

File: check_api/src/main/java/com/google/errorprone/FileReporter.java (hunk 3 above)
Comment (tbroyer) [reply check_api/src/main/java/com/google/errorprone/FileReporter.java]: How about using an `XMLStreamWriter`? https://hsivonen.fi/producing-xml/

File: check_api/src/main/java/com/google/errorprone/FileReporter.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public void started(final TaskEvent e) {
+    if (e.getKind() == TaskEvent.Kind.COMPILATION) {
+      if (pw != null) {
+        pw.println("<?xml version=\"1.0\" encoding=\"utf-8\"?>");
+        pw.println("<errorprone timestamp=\"" + ZonedDateTime.now() + "\">");
```
Comment (davidhart82) [inline check_api/src/main/java/com/google/errorprone/FileReporter.java]: This should be formatted, ideally with an ISO standard.

Comment (tbroyer) [reply check_api/src/main/java/com/google/errorprone/FileReporter.java]: It is already: `ZonedDateTime#toString()` produces an ISO 8601 timestamp.

Comment (davidhart82) [reply check_api/src/main/java/com/google/errorprone/FileReporter.java]: Relying on a `toString` implementation feels a bit fragile?  I might have used a `DateFormat` with the format specifier `"yyyy-MM-dd'T'HH:mm:ss.SSS"` just to make it more obvious?

**Comment by davidhart82** [comment]
How about having the report configured like so: `-XepReportFile:<formatType>:<path>`?

File: check_api/src/main/java/com/google/errorprone/JavacErrorDescriptionListener.java [hunk 6]
```diff
@@ -106,6 +110,7 @@ public void onDescribed(Description description) {
           break;
       }
       log.report(factory.create(type, log.currentSource(), pos, MESSAGE_BUNDLE_KEY, message));
+      fileReporter.report(type, log.currentSource(), pos, message);
```
Comment (davidhart82) [inline check_api/src/main/java/com/google/errorprone/JavacErrorDescriptionListener.java]: I wonder whether we should have a composite reporter, to abstract the use of both of these `report` methods?  Taking a `JCDiagnostic` type as the param to `report` does not seem a terrible idea either.

Comment (davidhart82) [reply check_api/src/main/java/com/google/errorprone/JavacErrorDescriptionListener.java]: Though I haven't looked at where `Log` manifests from!

**Comment by stalar** [comment]
I no longer have an interest in this. Feel free to continue working on the feature - use my code as you wish!

**Pull request closed without merging**